torchdata/dataloader2/reading_service.py

```diff
@@ -184,67 +181,57 @@ class PrototypeMultiProcessingReadingService(ReadingServiceInterface):
     datapipes: List
     end_datapipe: Optional[DataPipe]
     _mp: bool
     _pg: Optional[dist.ProcessGroup]
-    _world_size: int
-    _rank: int
+    _dist_info: DistInfo
 
     def __init__(
         self,
         num_workers: int = 0,
-        multiprocessing_context=None,
+        multiprocessing_context: Optional[str] = None,
         prefetch_worker: int = 10,
         prefetch_mainloop: int = 10,
+        worker_init_fn: Optional[Callable[[DataPipe, DistInfo, WorkerInfo], DataPipe]] = None,
+        worker_reset_fn: Optional[Callable[[DataPipe, DistInfo, WorkerInfo], DataPipe]] = None,
     ) -> None:
         self.num_workers = num_workers
-        # TODO(613): Should be one of 'fork', 'spawn'
+        if multiprocessing_context is not None:
+            _all_start_methods = mp.get_all_start_methods()
+            assert (
+                multiprocessing_context in _all_start_methods
+            ), f"Please choose one available multiprocessing context from {_all_start_methods}"
         self.multiprocessing_context = multiprocessing_context
         self.prefetch_worker = prefetch_worker
         self.prefetch_mainloop = prefetch_mainloop
+        self.worker_init_fn = worker_init_fn
+        self.worker_reset_fn = worker_reset_fn
         self.processes = []
         self.datapipes = []
         self.end_datapipe = None
         self._mp = num_workers > 0
         self._pg = None
-        self._world_size = 1
-        self._rank = 0
-
-    @staticmethod
-    def _process_init_fn(world_size, rank, num_workers, worker_id, datapipe):
-        global_worker_id = worker_id * world_size + rank
-        total_num_workers = num_workers * world_size
-        torch.utils.data.graph_settings.apply_sharding(datapipe, total_num_workers, global_worker_id)
-
-    @staticmethod
-    def _process_reset_fn(world_size, rank, num_workers, worker_id, datapipe, shared_seed):
-        # This function will receive worker local copy of datapipe and args value from ``initialize_iteration``
-        worker_seed_generator = torch.Generator()
-        worker_seed_generator.manual_seed(shared_seed)
-        torch.utils.data.graph_settings.apply_random_seed(
-            datapipe,
-            worker_seed_generator,
-        )
-        # Set different seeds across distributed workers
-        global_worker_id = worker_id * world_size + rank
-        worker_seed_generator.manual_seed(shared_seed + global_worker_id)
-
-        py_seed = _generate_random_seed(worker_seed_generator).item()
-        random.seed(py_seed)
-
-        torch_seed = _generate_random_seed(worker_seed_generator).item()
-        torch.manual_seed(torch_seed)
-
-        if HAS_NUMPY:
-            # Numpy only accepts uint32 as the seed
-            np_seed = _generate_random_seed(worker_seed_generator, torch.int32).item()
-            if np_seed < 0:
-                np_seed = 2 ** 32 + np_seed
-            numpy.random.seed(np_seed)
+        self._dist_info = DistInfo(1, 0)
```

Comment:
do we plan to support `world_size > 1` in the future? (understand it's a prototype now~)

Reply:
It's supporting `world_size > 1`. And, in some cases like `lightning`, they initialize distributed environment after `DataLoader` is constructed. So, I have to make sure `distributed` call moved to `__iter__`.